pkg/kubelet/cm/dra/manager.go

```diff
@@ -98,48 +101,82 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
-				}
-
-				response, err := client.NodePrepareResource(
-					context.Background(),
-					resourceClaim.Namespace,
-					resourceClaim.UID,
-					resourceClaim.Name,
-					resourceClaim.Status.Allocation.ResourceHandle)
-				if err != nil {
-					return fmt.Errorf("NodePrepareResource failed, claim UID: %s, claim name: %s, resource handle: %s, err: %+v",
-						resourceClaim.UID, resourceClaim.Name, resourceClaim.Status.Allocation.ResourceHandle, err)
+				// Build a slice of "resourceHandles" to group the name of the
+				// kubelet plugin to call NodePrepareResources() on and the
+				// ResourceHandle data to be processed by that plugin. For now
+				// this slice will only have a single entry, where the name of
+				// the kubelet plugin matches the DriverName. In the future we
+				// plan to allow each claim to be processed by multiple plugins
+				// (each with their own ResourceHandle) so this code is being
+				// written in a way to accommodate this.
+				resourceHandles := []struct {
+					KubeletPluginName string
+					Data              string
+				}{
+					{
+						KubeletPluginName: resourceClaim.Status.DriverName,
+						Data:              resourceClaim.Status.Allocation.ResourceHandle,
+					},
 				}
 
-				klog.V(3).InfoS("NodePrepareResource succeeded", "response", response)
-
-				// TODO: We are adding the claimInfo struct to the cache and syncing it to the checkpoint *after* the NodePrepareResource
-				// call has completed. This will cause issues if the kubelet gets restarted between NodePrepareResource and syncToCheckpoint.
-				// It will result in not calling NodeUnprepareResource for this claim because no claimInfo will be synced back to the cache
-				// for it after the restart. We need to resolve this issue before moving to beta.
-				claimInfo, err := newClaimInfo(
-					driverName,
+				// Create a claimInfo object to store the relevant claim info.
+				claimInfo := newClaimInfo(
+					resourceClaim.Status.DriverName,
 					resourceClaim.UID,
 					resourceClaim.Name,
 					resourceClaim.Namespace,
 					sets.New(string(pod.UID)),
-					response.CdiDevices)
-				if err != nil {
-					return fmt.Errorf("newClaimInfo failed, claim UID: %s, claim name: %s, claim namespace: %s, err: %+v",
-						resourceClaim.UID, resourceClaim.Name, resourceClaim.Namespace, err)
-				}
-
-				m.cache.add(claimInfo)
-				// Checkpoint to reduce redundant calls to NodePrepareResource() after a kubelet restart.
-				err = m.cache.syncToCheckpoint()
-				if err != nil {
-					return fmt.Errorf("failed to checkpoint claimInfo state, err: %+v", err)
+				)
+
+				// Walk through each resourceHandle
+				for _, resourceHandle := range resourceHandles {
+					// If no KubeletPluginName is provided, we use the DriverName
+					pluginName := resourceHandle.KubeletPluginName
+					if pluginName == "" {
+						pluginName = resourceClaim.Status.DriverName
+					}
```

Comment:
why do we need this check if you already put it in the resourceHandles  struct here [1]

[1]- https://github.com/kubernetes/kubernetes/pull/116513/commits/9702b0a464f1eb782ad94e9cd5fc6d3b8e73872b#diff-203d2de4519315883638490fcafedbe368e5d2dee0cdba88c91cef46614f80acR117

Reply:
I could remove it in this PR, but it will be necessary in the future, once we get the list of Resourcehandles directly from the AllocationResult.

Reply:
Updated